core/engine/src/value/conversions/try_from_js/collections.rs

```diff
@@ -62,3 +95,44 @@ where
             .collect()
     }
 }
+
+fn for_each_elem_in_js_map<F>(js_map: &JsMap, mut f: F, context: &mut Context) -> JsResult<()>
+where
+    F: FnMut(JsValue, JsValue, &mut Context) -> JsResult<()>,
+{
+    let unexp_obj_err = || {
+        JsResult::Err(
+            JsNativeError::typ()
+                .with_message("MapIterator return unexpected object")
+                .into(),
+        )
+    };
+
+    let iter = js_map.entries(context)?;
+    loop {
+        let next = iter.next(context)?;
+        let Some(iter_obj) = next.as_object() else {
+            return unexp_obj_err();
+        };
+
+        let done = iter_obj.get(js_str!("done"), context)?;
+        let Some(done) = done.as_boolean() else {
+            return unexp_obj_err();
+        };
+        if done {
+            break;
+        }
```

Comment:
Here I think you can use [`IteratorResult`](https://github.com/boa-dev/boa/blob/main/core/engine/src/builtins/iterable/mod.rs#L314) to do all the checks for you. We'll probably want to make `next` return that instead of the raw `JsValue`, but constructing it manually should be fine for now.